Implement the workflow pipeline

xiki-tempula opened the pull request:
I will try to Implement the workflow pipeline as is discussed in #111 .

I noticed that the subsampling part will be refactored with #98. Given I cannot leave this part blank, I have made a minimal implementation of the corresponding option and I'm happy to change it when #98 is merged.

All the information is logged using logging #34.

xiki-tempula commented:
@orbeckst I think this PR is ready for review. 
So I will sort the energy file according to the n_uk first and if n_uk is not present, it will fall back to sort via dhdl.
However, I don't have any test file for that and this part is not covered by test. Do you think I shall just omit this part for code coverage? Thank you.

orbeckst commented on src/alchemlyb/constants.py:
```diff
@@ -0,0 +1,5 @@
+'''Physical and mathematical constants and units.'''
+# Taken from scipy.constants since py2 doesn't support it
+k = 1.380649e-23
+N_A = 6.02214076e+23
```
Take constants from NIST and document each. Eg
```python
#: Boltzmann's constant :math:`k_B` in kJ/(mol K); value from `NIST CODATA: k<https://physics.nist.gov/cgi-bin/cuu/Value?k>`_.
k_Boltzmann = 1.380649e-23
```
Also use expressive variable names.

orbeckst commented on docs/visualisation.rst:
```diff
@@ -12,6 +12,7 @@ visualisation tools to help user to judge the estimate.
     plot_mbar_overlap_matrix
     plot_ti_dhdl
     plot_dF_state
+    plot_convergence
```
move plotting to separate PR

orbeckst commented on docs/visualisation.rst:
```diff
@@ -131,6 +132,53 @@ Will give a plot looks like this
    A bar plot of the free energy differences evaluated between pairs of adjacent
    states via several methods, with corresponding error estimates for each method.
 
+.. _plot_convergence:
```
move plotting to separate PR

orbeckst commented on docs/visualisation/alchemlyb.visualisation.plot_convergence.rst:
```diff
@@ -0,0 +1,19 @@
+.. _visualisation_plot_convergence:
```
move plotting to separate PR

orbeckst commented on docs/visualisation/alchemlyb.visualisation.plot_ti_dhdl.rst:
```diff
@@ -11,7 +11,7 @@ When custom labels are desirable, the user could pass a list of strings to the
 each alchemical transformation could also be set by passing a list of color
 string to the *colors*. The unit in the y axis could be labelled to other units
-by setting *units*, which by default is kcal/mol. The user can pass
+by setting *units*, which by default is kBT. The user can pass
```
Format kBT as LaTeX.

orbeckst commented on src/alchemlyb/estimators/ti_.py:
```diff
@@ -105,3 +105,46 @@ def fit(self, dHdl):
 
         return self
+
+    def separate_dhdl(self):
```
possibly separate PR + tests?

Is it only used for the workflow or more general?

xiki-tempula replied:
This is actually used by the plot convergence. I will move this and the plotting to another PR

orbeckst reviewed (changes requested):
This is a lot of stuff and adds a whole new thing (workflows) to alchemlyb. It looks interesting but this is too much to review in one piece.

Could you please break up the PR into small, independent pieces, eg
- visualization
- separate_dhdl() functionality
- workflow

This will make reviewing easier.

Introducing workflows is its own discussion, which will best be done on a dedicated PR. Maybe @dotsdl already has an opinion on this, too?

xiki-tempula commented:
@orbeckst Thanks for the review. I have move the separate dhdl and plot convergence to separate PRs #120 #121. The test will run as soon as those are merged.

xiki-tempula commented:
@orbeckst Thanks for the review. I have merged the master and ensure that this PR has the right coverage and test and have changed the kBT to kT.

orbeckst commented:
I've been pretty busy all around. I'd like to make a 0.4.0 release #113 with all the good stuff and then we can revisit workflows.

xiki-tempula commented:
@orbeckst Thanks for the work. I think this PR is ready for review now. Please take your time.

orbeckst commented on docs/workflow.rst:
```diff
@@ -0,0 +1,109 @@
+==================
+Though **alchemlyb** is a library offering great flexibility in deriving free
+energy estimate, it also provide a easy pipeline that is similar to
+`Alchemical Analysis <https://github.com/MobleyLab/alchemical-analysis>`_ and a
```
Use a reST link with alternative name
```reST
`Alchemical Analysis `_ and a ...


.. _Alchemical Analysis: https://github.com/MobleyLab/alchemical-analysis
```
especially as it is used multiple times. Btw, you can use 
```reST
The `displayed text 1 <link-target>`_ and `other text <link-target>`_ will link to the same
thing.

.. _`link-target`: https://example.com
```

orbeckst commented on docs/index.rst:
```diff
@@ -66,6 +66,7 @@ Contributions are very welcome. If you have bug reports or feature requests or q
     :caption: User Documentation
 
     install
+    workflow
```
workflows is not feature-complete yet so move it towards the end – otherwise people will read it first and expect that it already solves their problems

orbeckst commented on docs/workflow.rst:
```diff
@@ -0,0 +1,109 @@
+Automatic workflow
+==================
```
add a note that workflows is experimental

orbeckst commented on src/alchemlyb/tests/test_workflow.py:
```diff
@@ -0,0 +1,319 @@
+import numpy as np
```
If we get more workflows and each is a separate module, make a workflows directory and create one `workflows/test_WORKFLOWNAME.py` for each.

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,700 @@
+
+
+class ABFE():
+    '''Alchemical Analysis style automatic workflow.
```
The documentation does not make clear if this is a general workflow class or specific for one type of free energy calculations.

xiki-tempula replied:
Sorry. I use this workflow for my ABFE and RBFE calculations and it will work for these two types of calculations but I think the general workflow should be similar for other types of calculations?

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,700 @@
+    dHdl_list : list
+        The list of dHdl read from the files.
+    '''
+    def __init__(self, units='kcal/mol', software='Gromacs', dir='./',
```
for `dir` use `dir=os.path.curdir` or something suitably OS-independent.

orbeckst replied:
Have units default to kT instead of kcal/mol to be consistent with the rest of alchemlyb.

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,700 @@
+        points (an integer) must be provided. Default: None. (not doing
+        convergence analysis).
+    log : str
+        The filename of the log file. Default: 'result.log'
```
mention that the logger logs to "alchemlyb.workflows.ABFE"

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,700 @@
+
+        logging.basicConfig(filename=log, level=logging.INFO)
+        self.logger = logging.getLogger('alchemlyb.workflows.ABFE')
+        self.logger.info('Initialise Alchemlyb ABFE Workflow')
```
add version information `alchemlyb.__version__`

orbeckst reviewed (changes requested):
This looks like useful but specialized code. It will work for GROMACS ABFE simulations but nothing else. It's quite a lot of code but you wrote a huge amount of tests to cover it. It looks like a good example how one achieve more complicated solutions with the alchemlyb building blocks.

My main concern is that it is not clear how the workflows module should grow in the future. The PR does not define what an alchemlyb workflow ought to be. How would we decide in the future what we should or should not add to the workflows module? What standards would we measure new contributions by? 

If we were to add this PR, it would be a departure from the [current philosophy & principles](https://alchemlyb.readthedocs.io/en/latest/api_principles.html) 

> It includes functions for parsing data from formats common to existing MD engines, subsampling these data, and fitting these data with an estimator to obtain free energies.
> These functions are simple in usage and pure in scope, and can be chained together to build customized analyses of data.
> 
> alchemlyb seeks to be as boring and simple as possible to enable more complex work. 

In the short term it could be convenient to have complex workflows directly available. I worry that over the long term it makes it more difficult to maintain the library. 

Nevertheless, we could decide that we want to broaden alchemlyb. In this case I would, however, like to know the *functionality* that the `workflows` module should ultimately contain (i.e., a list of problems for which workflows  must be added to make it usable for most users), how these workflows are structured – *definition of a common API*, what *building blocks for workflows* should be added to reduce boiler plate code, and *criteria for including workflows*. 

Finally, why do the workflows have to be inside alchemlyb? Couldn't they be a package built with alchemlyb as dependency? `pip install alchemlyb_workflows` seems easy enough and provides a lot more freedoms.

@xiki-tempula can you elaborate a bit more and answer the questions I posed above? 
 
@davidlmobley  @dotsdl @mrshirts I'd be keen to also hear your opinion.

xiki-tempula commented:
@harlor Thank you very much for the code review. I think the goal of the workflow is to promote best practices and more importantly to provide a simple example that the user could follow so I would like to trade abstraction for clarity.

However, it is still worth hearing opinions about how different blocks of alchemlyb could be improved.

parsing

I do agree that one could do 

```
from alchemlyb.parsing import get_extract_u_nk
extract_u_nk = get_extract_u_nk('gmx')
```

so as to avoid direct mapping.

I kind of leave it like this as I kind of need to do

```
if engine.lower() == 'gromacs':
    extract_u_nk = get_extract_u_nk('gmx')
```

anyway. So I think the extra layer of wrapper could be added but the current implementation should still be ok.

decorrelation

As is discussed in #159, I was originally waiting for #98 but I will try to implement an easy one first.

estimating

I think the current level of abstraction in `alchemlyb.estimators` is good enough.

plotting

I think the current level of abstraction in `alchemlyb.visualisation ` is good enough.

output (text, JSON)

This is a difficult thing. I do agree that this could be abstracted but I fear that it might be too focused.

@orbeckst Do you think it is worth creating a function (alchemlyb.postprocessors.text_output), which takes in a list of estimators and return a string that has alchmeical-analysis type formating? Or it is too specific to ABFE (since the main selling point would be to separate out bonded coul and vdw)?

CLI

I think you have written excellent code in flamel so when this PR is merged, I will just use your interface.

Thanks again for all the suggestions.

xiki-tempula commented:
@orbeckst Thank you very much for the review. I wonder if you mind giving a review of this, please? Thank you.

orbeckst commented:
@xiki-tempula I am swamped this week and don't have time for big PR reviews. Can you please ping me again Monday next week? Sorry.

Perhaps someone else has time to comment? @alchemistry/alchemlyb ?

xiki-tempula commented:
@orbeckst Would you mind having a look at this PR and giving me some advice, please? Thank you.

orbeckst commented on src/alchemlyb/tests/test_workflow_ABFE.py:
```diff
@@ -0,0 +1,333 @@
+        dHdl_sample_list = workflow.dHdl_sample_list
+        delattr(workflow, 'dHdl_sample_list')
+        workflow.check_convergence(10, estimator='ti')
+        os.remove('dF_t.pdf')
```
handle cleanup elsewhere (or rather, use a tmpdir for output)

orbeckst commented on src/alchemlyb/tests/test_workflow_ABFE.py:
```diff
@@ -0,0 +1,333 @@
+    def test_convergence_dhdl(self, workflow):
+        '''test if the dF_state.pdf has been plotted.'''
+        workflow.check_convergence(10, estimator='ti')
+        os.remove('dF_t.pdf')
```
handle cleanup elsewhere (or rather, use a tmpdir for output)

orbeckst commented on src/alchemlyb/tests/test_workflow_ABFE.py:
```diff
@@ -0,0 +1,333 @@
+        '''test if the dF_state.pdf has been plotted.'''
+        workflow.check_convergence(10, estimator='ti', dF_t='dF_t.pdf')
+        assert os.path.isfile('dF_t.pdf')
+        os.remove('dF_t.pdf')
```
handle cleanup elsewhere (or rather, use a tmpdir for output)

orbeckst commented on src/alchemlyb/tests/test_workflow_ABFE.py:
```diff
@@ -0,0 +1,333 @@
+    def test_O_MBAR(self, workflow):
+        '''test if the O_MBAR.pdf has been plotted.'''
+        assert os.path.isfile('O_MBAR.pdf')
+        os.remove('O_MBAR.pdf')
```
handle cleanup elsewhere (or rather, use a tmpdir for output) ... and for the others, too

orbeckst commented on src/alchemlyb/tests/test_workflow_ABFE.py:
```diff
@@ -0,0 +1,333 @@
+        workflow = ABFE(units='kcal/mol', software='Gromacs', dir=dir,
+                        prefix='dhdl', suffix='bz2', T=310, skiptime=0,
+                        uncorr='dhdl', threshold=50,
+                        methods=('mbar', 'bar', 'ti'), out='./',
```
use tmpdir for out

orbeckst commented on src/alchemlyb/tests/test_workflow_ABFE.py:
```diff
@@ -0,0 +1,333 @@
+    def test_dhdl_TI_noTI(self, workflow):
+        '''Test to plot the dhdl_TI when ti estimator is not there'''
+        full_estimator = workflow.estimator
+        workflow.estimator.pop('ti')
```
unsafe, need function level fixture or mock/patch

orbeckst commented on src/alchemlyb/tests/test_workflow_ABFE.py:
```diff
@@ -0,0 +1,333 @@
+        os.remove('dF_t.pdf')
+        assert len(workflow.convergence) == 10
+
+    def test_convergence_TI(self, workflow):
```
"plot" or better, also test numerical values

orbeckst commented on src/alchemlyb/tests/test_workflow_ABFE.py:
```diff
@@ -0,0 +1,333 @@
+        assert len(workflow.convergence) == 10
+        workflow.dHdl_sample_list = dHdl_sample_list
+
+    def test_convergence_dhdl(self, workflow):
```
"plot" or better, also test numerical values

orbeckst commented on src/alchemlyb/tests/test_workflow_ABFE.py:
```diff
@@ -0,0 +1,333 @@
+        workflow.u_nk_sample_list = u_nk_sample_list
+
+    def test_convergence_nosample_dhdl(self, workflow):
+        '''test if the dF_state.pdf has been plotted.'''
```
"plot" or better, also test numerical values

orbeckst commented on src/alchemlyb/tests/test_workflow_ABFE.py:
```diff
@@ -0,0 +1,333 @@
+        os.remove('dF_t.pdf')
+        assert len(workflow.convergence) == 10
+
+    def test_convergence_nosample_u_nk(self, workflow):
```
"plot" or better, also test numerical values

orbeckst commented on src/alchemlyb/tests/test_workflow_ABFE.py:
```diff
@@ -0,0 +1,333 @@
+        full_estimator = workflow.estimator
+        workflow.estimator.pop('ti')
+        workflow.plot_ti_dhdl(dhdl_TI='dhdl_TI.pdf')
+        assert os.path.isfile('dhdl_TI.pdf') == False
```
assert not isfile(....)

orbeckst commented on src/alchemlyb/tests/test_workflow_ABFE.py:
```diff
@@ -0,0 +1,333 @@
+                        methods=('mbar', 'bar', 'ti'), out='./',
+                        overlap='O_MBAR.pdf',
+                        breakdown=True, forwrev=10, log='result.log')
+        return workflow
```
actually, yield is not needed when using tmpdir

orbeckst commented on src/alchemlyb/tests/test_workflow_ABFE.py:
```diff
@@ -0,0 +1,333 @@
+    def test_O_MBAR(self, workflow):
+        '''test if the O_MBAR.pdf has been plotted.'''
+        assert os.path.isfile('O_MBAR.pdf')
+        os.remove('O_MBAR.pdf')
```
actually yield is not needed when using tmpdir, so definitely use tmpdir or the tempdir factory

orbeckst commented on src/alchemlyb/tests/test_workflow_ABFE.py:
```diff
@@ -0,0 +1,333 @@
+        workflow.plot_ti_dhdl(dhdl_TI='dhdl_TI.pdf')
+        workflow.plot_dF_state(dF_state='dF_state.pdf')
+        workflow.check_convergence(10, dF_t='dF_t.pdf')
+        return workflow
```
actually yield is not needed when using tmpdir, so definitely use tmpdir or the tempdir factory

orbeckst commented on src/alchemlyb/tests/test_workflow_ABFE.py:
```diff
@@ -0,0 +1,333 @@
+        assert all([len(u_nk) == 4001 for u_nk in workflow.u_nk_list])
+        assert all([len(dHdl) == 4001 for dHdl in workflow.dHdl_list])
+
+    def test_estomator(self, workflow):
```
estimator

orbeckst commented on src/alchemlyb/tests/test_workflow_ABFE.py:
```diff
@@ -0,0 +1,333 @@
+        os.remove('dF_t.pdf')
+        assert len(workflow.convergence) == 10
+
+class Test_unpertubed_lambda():
```
Same comments above about using tmpdir.

orbeckst commented on src/alchemlyb/tests/test_workflow_ABFE.py:
```diff
@@ -0,0 +1,333 @@
+    '''Test the if two lamdas present and one of them is not pertubed.
+
+                                       fep  bound
+time    fep-lambda bound-lambda
```
indent documentation

orbeckst commented on src/alchemlyb/tests/test_workflow_ABFE.py:
```diff
@@ -0,0 +1,333 @@
+        workflow.update_units('kcal/mol')
+        workflow.preprocess(skiptime=10, uncorr='dhdl', threshold=50)
+        workflow.estimate(methods=('mbar', 'bar', 'ti'))
+        workflow.plot_overlap_matrix(overlap='O_MBAR.pdf')
+        workflow.plot_ti_dhdl(dhdl_TI='dhdl_TI.pdf')
+        workflow.plot_dF_state(dF_state='dF_state.pdf')
+        workflow.check_convergence(10, dF_t='dF_t.pdf')
```
These could be run in the test methods when needed.

xiki-tempula replied:
@orbeckst So the aim of this fixture is to mimic running the whole process and check if everything works ok.

orbeckst replied:
ok

orbeckst commented on src/alchemlyb/tests/test_workflow_ABFE.py:
```diff
@@ -0,0 +1,333 @@
+            dHdl.set_index('bound-lambda', append=True, inplace=True)
+
+        workflow.estimate(methods=('ti', ))
+        workflow.plot_ti_dhdl(dhdl_TI='dhdl_TI.pdf')
+        workflow.plot_dF_state(dF_state='dF_state.pdf')
+        workflow.check_convergence(10, dF_t='dF_t.pdf', estimator='ti')
```
These could be run in the test methods when needed.

orbeckst commented on src/alchemlyb/tests/test_workflow_ABFE.py:
```diff
@@ -0,0 +1,333 @@
+        os.remove('dF_t.pdf')
+        assert len(workflow.convergence) == 10
+
+    def test_single_estimator_mbar(self, workflow):
```
mbar -> ti ?

orbeckst commented on src/alchemlyb/tests/test_workflow_ABFE.py:
```diff
@@ -0,0 +1,333 @@
+        summary = workflow.generate_result()
+        assert np.isclose(summary['TI']['Stages']['TOTAL'], 2.946, 0.1)
+
+class Test_methods():
```
same comments as above re: tmpdir etc

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,651 @@
+        The unit used for printing and plotting results. {'kcal/mol', 'kJ/mol',
+        'kT'}. Default: 'kT'.
+    software : str
+        The software used for generating input. {'Gromacs', }
```
I think GROMACS is capitalized.

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,651 @@
+                 overlap=None, breakdown=None, forwrev=None,
+                 log='result.log'):
+
+        logging.basicConfig(filename=log, level=logging.INFO)
```
I don't think that you should override the logging config that may be set by calling code. Remove the config.

Users need to configure their logger. You can add a `alchemlyb.start_logging()` function (see `MDAnalysis.start_logging()` and add docs. alchemlyb is a library first, it does not have to do everything for the user. flamel can set its logger.

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,651 @@
+        logging.basicConfig(filename=log, level=logging.INFO)
+        self.logger = logging.getLogger('alchemlyb.workflows.ABFE')
+        self.logger.info('Initialise Alchemlyb ABFE Workflow')
+        self.logger.info('Alchemlyb Version: {}'.format(__version__))
```
Use standard logger variable interpolation
```
self.logger.info('Alchemlyb Version: %s', __version__)
```
with % interpolation.

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,651 @@
+        self.logger.info('Initialise Alchemlyb ABFE Workflow')
+        self.logger.info('Alchemlyb Version: {}'.format(__version__))
+
+        self.logger.info('Set temperature to {} K.'.format(T))
```
%

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,651 @@
+
+        self.update_units(units)
+
+        self.logger.info('Finding files with prefix: {}, suffix: {} under '
+                         'directory {} produced by {}'.format(prefix, suffix,
+                                                              dir, software))
```
%

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,651 @@
+                software))
+            extract_u_nk = gmx.extract_u_nk
+            extract_dHdl = gmx.extract_dHdl
+        else: # pragma: no cover
```
I want to see that this workflow can really work for other inputs, too.

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,651 @@
+        analysis will not be performed at all. Default: 50.
+    methods : str
+        A list of the methods to esitimate the free energy with. Default: None.
+    out : str
```
rename to outdirectory or something more expressive

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,651 @@
+    out : str
+        Directory in which the output files produced by this script will be
+        stored. Default: os.path.curdir.
+    overlap : str
```
Why here? Can't this be just an arg for the plotting method?

xiki-tempula replied:
This is now moved to run().

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,651 @@
+    overlap : str
+        The filename for the plot of overlap matrix. Default: None. (not
+        plotting).
+    breakdown : bool
```
Does this have to be here? Can't this be controlled through a separate method?

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,651 @@
+        with the specified number of points in the time plot. The number of time
+        points (an integer) must be provided. Default: None. (not doing
+        convergence analysis).
+    log : str
```
This should be controlled via logger — just add a note on how to configure a logger.

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,651 @@
+                 prefix='dhdl', suffix='xvg', T=298, skiptime=0, uncorr=None,
+                 threshold=50, methods=None, out=os.path.curdir,
+                 overlap=None, breakdown=None, forwrev=None,
+                 log='result.log'):
```
I really dislike running everything in init. As soon as I create the object, I am stuck with running it.

Please refactor to make it more modular. init just sets variables. Then have methods for file loading, estimation, and plotting. 

Have a run() or execute() method to actually perform the workflow.

xiki-tempula commented:
@orbeckst Thanks for the comment. Sorry, I'm a bit busy before BPS 2021 and will work on this when that part is finished.

orbeckst commented:
Yes, of course. Sorry, I am currently snowed under deadlines and it is has been difficult to commit much time to anything else.

xiki-tempula commented:
@orbeckst I think the general structure could be

```
class AnalysisBase():
    def __init__(self, units='kT', software='Gromacs', T=298, *args, **kwargs):
        self.T = T
        self.software = software
        self.unit = units
        self.file_list = []
    
    def read(self, *args, **kwargs):
        self.u_nk_list = []
        self.dHdl_list = []
    
    def subsample(self, *args, **kwargs):
        self.dHdl_sample_list = []
        self.u_nk_sample_list = []
    
    def estimate(self, *args, **kwargs):
        self.result = pd.DataFrame()
    
    def convergence(self, *args, **kwargs):
        self.convergence = pd.DataFrame()
    
    def plot(self, *args, **kwargs):
        ...
```

orbeckst commented:
That's an interesting idea. How will a user use the class? Can you show mock-code that shows how it would be used in practice?

xiki-tempula commented:
@orbeckst For full automatic analysis, the user run
```
wf = AnalysisBase()
wf.run()
```
Which would read the data, subsample the data, do the estimate, plot and convergence analysis.
The analysis could be done step by step
```
wf = AnalysisBase()
wf.read()
wf.subsample()
wf.estimate()
wf.convergence()
wf.plot()
```

The convergence and plot could be ignored.
The analysis could be done step by step
```
wf = AnalysisBase()
wf.read()
wf.subsample()
wf.estimate()
```

Running
```
wf = AnalysisBase()
wf.read()
wf.estimate()
```
would issue a warning and use u_nk_list and dHdl_list instead of dHdl_sample_list and u_nk_sample_list.

Running
```
wf = AnalysisBase()
wf.convergence()
wf.plot()
```
Would give an error of needing to run the read first.

orbeckst commented:
This looks quite clean (and I obviously like the AnalysisBase approach from MDAnalysis.analysis ;-)). 

Presumably, we would have a `workflows.base.AnalysisBase` class and then derive, e.g.,
```python
class ABFEAnalysis(base.AnalysisBase):
   ...
```

I would perhaps actually name the base class `base.Workflow` or `base.WorkflowBase` instead of AnalysisBase to avoid any confusion with the MDAnalysis base class.

Overall, I do like your approach. It seems to have enough flexibility for future growth but also contains a common structure.

xiki-tempula commented on docs/workflow.rst:
```diff
@@ -0,0 +1,134 @@
+.. currentmodule:: alchemlyb.workflows
+
+.. autoclass:: ABFE
+   :noindex:
```
@orbeckst I would like to explain the functions in order.

Explain the class
auto-generated class explanation
Explain the class.function_1
auto-generated class.function_1 explanation
Explain the class.function_2
auto-generated class.function_2 explanation

xiki-tempula commented:
@orbeckst With regard to the logger. I wonder if you feel that it might be better to use f-sting? I saw some repo using it. So we can have the new style string formating in the logger.

xiki-tempula commented:
@orbeckst Thank you for the help. I think this PR is ready for review now.

orbeckst commented:
@xiki-tempula I won’t be able to look at it over the next week, please be patient. Thanks.

orbeckst commented on docs/workflows/alchemlyb.workflows.ABFE.rst:
```diff
@@ -0,0 +1,96 @@
+           TOTAL     35.137291    0.103580  34.860619   0.087022  34.924618  0.119206
+
+The :ref:`overlay matrix for the MBAR estimator <plot_overlap_matrix>` will be
+plotted and saved to `O_MBAR.pdf`.
```
Use reST markup, [sphinx file role]*https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#role-file):
```suggestion
plotted and saved to :file:`O_MBAR.pdf`.
```

Only use single backticks for referencing parameters from the _Parameters_ list.

orbeckst commented on docs/workflows/alchemlyb.workflows.ABFE.rst:
```diff
@@ -0,0 +1,96 @@
+The :ref:`overlay matrix for the MBAR estimator <plot_overlap_matrix>` will be
+plotted and saved to `O_MBAR.pdf`.
+
+The :ref:`dHdl for TI <plot_TI_dhdl>` will be plotted to `dhdl_TI.pdf`.
```
use file role

orbeckst commented on docs/workflows/alchemlyb.workflows.ABFE.rst:
```diff
@@ -0,0 +1,96 @@
+
+The :ref:`dHdl for TI <plot_TI_dhdl>` will be plotted to `dhdl_TI.pdf`.
+
+The :ref:`dF states <plot_dF_states>` will be plotted to `dF_state.pdf` in
```
use file role

orbeckst commented on docs/workflows/alchemlyb.workflows.ABFE.rst:
```diff
@@ -0,0 +1,96 @@
+The :ref:`dHdl for TI <plot_TI_dhdl>` will be plotted to `dhdl_TI.pdf`.
+
+The :ref:`dF states <plot_dF_states>` will be plotted to `dF_state.pdf` in
+portrait model and `dF_state_long.pdf` in landscape model.
```
use file role

orbeckst commented on docs/workflows/alchemlyb.workflows.ABFE.rst:
```diff
@@ -0,0 +1,96 @@
+The :ref:`dF states <plot_dF_states>` will be plotted to `dF_state.pdf` in
+portrait model and `dF_state_long.pdf` in landscape model.
+
+The forward and backward convergence will be plotted to `dF_t.pdf` using
```
use file role

orbeckst commented on docs/workflows/alchemlyb.workflows.ABFE.rst:
```diff
@@ -0,0 +1,96 @@
+    >>>                 prefix='dhdl', suffix='xvg', T=298, out='./')
+    >>> workflow.run(skiptime=10, uncorr='dhdl', threshold=50,
+    >>>              methods=('mbar', 'bar', 'ti'), overlap='O_MBAR.pdf',
+    >>>              breakdown=True, forwrev=10)
```
Add a paragraph explaining the philosophy and assumptions behind this workflow. For example, unlike other parts of alchemlyb, it finds input files inside a directory. Tell the user that certain assumptions have to hold (which ones? eg file naming?). It also sets special names for output files. What is the recommended use for the workflow, eg, input and output into the same directory?

Basically, talk a bit more about the stuff that's obvious to you but not to someone who's just been using the other parts of alchemlyb. What should you understand when using the wf?

orbeckst commented on docs/workflows/alchemlyb.workflows.ABFE.rst:
```diff
@@ -0,0 +1,96 @@
+------------------------
+A interface similar to
+`Alchemical Analysis <https://github.com/MobleyLab/alchemical-analysis>`_
+could be excuted with two lines of command. ::
```
reference the actual ABFE class here somewhere

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,665 @@
+
+
+class ABFE(WorkflowBase):
+    '''Alchemical Analysis style automatic workflow.
```
```suggestion
    '''Workflow for absolute and relative binding free energy calculations.
    
    This workflow provides functionality similar to the ``alchemical-analysis.py`` script.
    It loads multiple input files from alchemical free energy calculations and computes the
    free energies between different alchemical windows using different estimators. It 
    produces plots to aid in the assessment of convergence.
    
```

You can add more and you can rewrite, of course, but add a concise summary. Possibly add links eg to alchemical-analysis.

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,665 @@
+        The unit used for printing and plotting results. {'kcal/mol', 'kJ/mol',
+        'kT'}. Default: 'kT'.
+    software : str
+        The software used for generating input. {'GROMACS', 'AMBER'}
```
State how this affects the workflow, e.g., used for choosing the file parser.

orbeckst replied:
State that `software` is case-insensitive.

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,665 @@
+    software : str
+        The software used for generating input. {'GROMACS', 'AMBER'}
+    dir : str
+        Directory in which data files are stored. Default: os.path.curdir.
```
If these are input files then make this clear. What do the files need to look like. Does it have to be a flat directory? How are the files found by the workflow? This is really important for anyone wanting to use it: what does the code expect to find?

orbeckst replied:
Somewhere you need to describe *how* inputfiles are detected and sorted, i.e., how are dir, prefix, suffix used.

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,665 @@
+        self.logger.info(f'Finding files with prefix: {prefix}, suffix: '
+                         f'{suffix} under directory {dir} produced by '
+                         f'{software}')
+        self.file_list = glob(dir + '/**/' + prefix + '*' + suffix,
+                         recursive=True)
```
Describe the file finding in the docs.

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,665 @@
+            self._extract_u_nk = amber.extract_u_nk
+            self._extract_dHdl = amber.extract_dHdl
+        else: # pragma: no cover
+            raise NameError(f'{software} parser not found.')
```
This is the wrong type of exception. It's a `ValueError` (wrong value).

Instead of "not found" say "not implemented".

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,665 @@
+        '''
+        u_nk_list = []
+        dHdl_list = []
+        for xvg in self.file_list:
```
Instead of `xvg` call this variable something more general, it's not just  xvg.

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,665 @@
+                u_nk_list.append(u_nk)
+            except: # pragma: no cover
+                self.logger.warning(
+                    f'Error reading read u_nk from {xvg}.')
```
Why not die here immediately? It looks like a recipe for wrong/incomplete results. What's the rationale of only issuing a warning?

I would prefer that the workflow stops unless one explicitly sets something like `ignore_warnings=True` or something.

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,665 @@
+                dHdl_list.append(dhdl)
+            except: # pragma: no cover
+                self.logger.warning(
+                    f'Error reading read dhdl from {xvg}.')
```
same questions as above

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,665 @@
+            index_list = sorted(range(len(self.file_list)),
+                key=lambda x:column_names.index(
+                    u_nk_list[x].reset_index('time').index.values[0]))
+        else:
+            self.logger.info('Sort files according to the dHdl.')
+            column_names = sorted([dHdl.reset_index('time').index.values[0]
+                                   for dHdl in dHdl_list])
+            index_list = sorted(range(len(self.file_list)),
+                key=lambda x:column_names.index(
+                    dHdl_list[x].reset_index('time').index.values[0]))
```
When is the `else` block L123 - L128 being executed? I am a bit worried when a substantial code block is not tested.

(Or is it only the logging line that's missing????)

xiki-tempula replied:
@orbeckst I have removed this code block

orbeckst commented on src/alchemlyb/tests/test_workflow_ABFE.py:
```diff
@@ -0,0 +1,303 @@
+        assert len(workflow.convergence) == 10
+
+class Test_automatic_amber():
+    '''Test the full automatic workflow for load_ABFE from alchemtest.gmx for
```
gmx??

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,665 @@
+            found to be less than this number. If 0 is given, the time series
+            analysis will not be performed at all. Default: 50.
+        methods : str
+            A list of the methods to esitimate the free energy with. Default:
```
```suggestion
            A list of the methods to estimate the free energy with. Default:
```

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,665 @@
+            Proceed with correlated samples if the number of uncorrelated samples is
+            found to be less than this number. If 0 is given, the time series
+            analysis will not be performed at all. Default: 50.
+        methods : str
```
type is list or tuple, not str

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,665 @@
+        ----------
+        skiptime : float
+            Discard data prior to this specified time as 'equilibration' data. Units
+            picoseconds. Default: 0.
```
How do you know that the input data are in ps? I think the only thing you can say here is that it's the same unit that was stored by the code that wrote it. alchemlyb does not really know about time.

xiki-tempula replied:
@orbeckst I think The u_nk pandas data frame has one column which is time?

orbeckst replied:
Yes, but we have no idea what the unit of this time column is. You don't need code changes, but remove "Units picoseconds" and replaces with "The unit of time is the same as in the output from the package that generated the time series of dH/dl and u_nk values."

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,665 @@
+            A list of the methods to esitimate the free energy with. Default:
+            `('mbar', 'bar', 'ti')`.
+        overlap : str
+            The filename for the plot of overlap matrix. Default: `O_MBAR.pdf`.
```
use file role or just a string

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,665 @@
+        breakdown : bool
+            Plot the free energy differences evaluated for each pair of adjacent
+            states for all methods, including the dH/dlambda curve for TI. Default:
+            True.
```
Use markup for True, False
```suggestion
            ``True``.
```

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,665 @@
+        self.u_nk_sample_list = None
+        self.dHdl_sample_list = None
+
+    def run(self, skiptime=0, uncorr='dhdl', threshold=50,
+            methods=('mbar', 'bar', 'ti'), overlap='O_MBAR.pdf',
+            breakdown=True, forwrev=10, *args, **kwargs):
```
Given that you have kwargs with defaults, you might not need to specify the defaults in the text. That makes it easier to keep everything consistent.

xiki-tempula replied:
@orbeckst Sorry, I wonder what do you mean by specify the defaults in the text. What is this test?

orbeckst replied:
In the docstring (=text) you often state the default for a kwarg. There's nothing wrong with doing that and you don't really need to change anything. I was only saying that *if* the defaults are set with the kwargs themselves then sphinx will generate documentation where the defaults are visible in the function signature and then it's not strictly necessary to manually write the defaults in the doc string. That has the advantage of not having to update the default in the docstring if the actual default value in the code ever gets changed.

It's a minor thing and as I said, you don't really need to change your docstring. If you _want_ to omit manually written defaults in the future then I'd be happy with that. That's all I really wanted to say.

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,665 @@
+        forwrev : int
+            Plot the free energy change as a function of time in both directions,
+            with the specified number of points in the time plot. The number of time
+            points (an integer) must be provided. Specify as `None` will not do
```
double backticks for ``None``

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,665 @@
+
+        Parameters
+        ----------
+        units : str
```
You can give the options directly
```suggestion
        units : {'kcal/mol', 'kJ/mol', 'kT'}
```

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,665 @@
+        ----------
+        units : str
+            The unit used for printing and plotting results. {'kcal/mol',
+            'kJ/mol', 'kT'}
```
... and then remove from here

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,665 @@
+            'kJ/mol', 'kT'}
+
+        '''
+        if units is not None:
+            self.logger.info(f'Set unit to {units}.')
+            self.units = units
+        else: # pragma: no cover
+            pass
```
Instead of this if statement, make `units=None` a kwarg.

xiki-tempula replied:
@orbeckst Sorry I wonder what do you mean?

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,665 @@
+
+
+    def update_units(self, units):
+        '''Update the unit.
```
what happens if `units == None`?

xiki-tempula replied:
@orbeckst Nothing will happen. The unit will not be modified.

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,665 @@
+            plt.close(ax.figure)
+
+
+    def update_units(self, units):
```
See other comment: make it optional `units=None` to simplify the code below.

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,665 @@
+        ----------
+        skiptime : float
+            Discard data prior to this specified time as 'equilibration' data.
+            Units picoseconds. Default: 0.
```
see other comment on time — I don't think we can guarantee that we have ps

xiki-tempula replied:
@orbeckst Both Gromacs and Amber would provide the time in the pandas dataframe generated by alchemlyb.parsing?

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,665 @@
+        uncorr : str
+            The observable to be used for the autocorrelation analysis; 'dhdl'
+            (obtained as a sum over those energy components that are changing).
+            Default: `dhdl`
```
It's a str so write it as "dhdl" (or leave out because kwarg defines in function signature)

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,665 @@
+
+        Parameters
+        ----------
+        methods : str
```
tuple or list or str

orbeckst reviewed (changes requested):
It's a big PR with an impressive amount of work so consider this a first round of comments. Sorry for the delay.

I have a bunch of comments on documentation; I think the important point is for you to put yourself into the shoes of someone who wants to use it and answer the obvious questions they have — ideally, you have someone else use the code with the current docs and then observe them to see if there's anything missing that they should know.

I also noted that there are many steps where only a warning is logged but the workflow continues. Is it normal to run the workflow with the expectation that some stuff won't work? Wouldn't it be safer and aid in reproducibility if the default was to fail immediately and require an explicit `ignore_warnings=True` or `force=True` or something like that to allow running an potentially incomplete workflow?

On that note, add to your example as an initial line to **enable the logger** — I don't think that alchemlyb enables logging by default, or do we??

xiki-tempula commented:
@orbeckst Thank you very much for the review. I have addressed some comments and have replied to the other comments which I might need further guidance.

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,675 @@
+        self.logger.info('Start convergence analysis.')
+        self.logger.info('Check data availability.')
+
+        if estimator.lower() in ['mbar', 'bar', 'autombar']:
```
Is there a an automatic way or at least a centralized way that we can keep this list? Our docs already differentiate between FEP-type and TI-type estimators. Maybe at the top level of `alchemlyb.estimators`?? Maybe
```python
# in estimators/__init__.py
...

FEP_ESTIMATORS = [MBAR.__name__, AutoMBAR.__name__, BAR.__name__]
TI_ESTIMATORS = [TI.__name__]
```

xiki-tempula replied:
Sorry, I see now that I need to add this in.

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,675 @@
+        matplotlib.axes.Axes
+            An axes with the convergence drawn.
+        '''
+        self.logger.info('Start convergence analysis.')
```
See other comments but if we already have lists like `estimators.FEP_ESTIMATORS` and `estimators.TI_ESTIMATORS` then we could check against the union of the two lists.

orbeckst replied:
Alternative idea (also see other comments): Do not add a check here but do below
```python
if estimator.lower() in FEP_ESTIMATORS:
   # do u_nk things
elif estimator.lower() in TI_ESTIMATORS:
   # do dHdl things
else:
  msg = ...
  logger.error(msg)
  raise ValueError(msg)
```

xiki-tempula replied:
Sorry, do we have estimators.FEP_ESTIMATORS? I don't think it is there?

```python
>>> from alchemlyb.estimators import FEP_ESTIMATORS
Traceback (most recent call last):
  File "<input>", line 1, in <module>
ImportError: cannot import name 'FEP_ESTIMATORS' from 'alchemlyb.estimators' (/Users/zhiyiwu/GitHub/alchemlyb/src/alchemlyb/estimators/__init__.py)
```

orbeckst reviewed (changes requested):
Great improvements, please see comments.

(Sorry, need to rush.)

xiki-tempula commented on docs/workflows/alchemlyb.workflows.ABFE.rst:
```diff
@@ -0,0 +1,130 @@
+The command would give the free energy estimate using all of
+:class:`~alchemlyb.estimators.TI`, :class:`~alchemlyb.estimators.BAR`,
+:class:`~alchemlyb.estimators.MBAR` and the result will be stored in
+:attr:`~alchemlyb.workflows.ABFE.summary` as :class:`pandas.Dataframe`. ::
```
The link does work. I have build it locally and :attr:`~alchemlyb.workflows.ABFE.summary` could give the correct link.

xiki-tempula commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,675 @@
+    dir : str
+        Directory in which data files are stored. Default: os.path.curdir.
+        The input files are searched using the pattern of
+        ``dir + '/**/' + prefix + '*' + suffix``.
```
Sorry, do you mean like ```dir` + '/**/' + `prefix` + '*' + `suffix```?
I think if I do `prefix`, I will need to remove the `` of the whole thing which might make it less clear?

xiki-tempula commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,675 @@
+
+        Parameters
+        ----------
+        forwrev : int
```
So the `forwrev` will also be a keyword in run(). 

I do agree that for check_convergence(), the keyword `n_timepoints` is more clear, but it will be a bit confusing if we put `n_timepoints` as a keyword in run(). 
It would also be confusing if we have two different keywords to mean the same thing in run() and check_convergence().

xiki-tempula commented:
@orbeckst Many thanks for the review. I have addressed the comments and have raised some questions about some of the comments.

orbeckst reviewed (changes requested):
Thank you for continuing your work on the PR. I found a few minor things (e.g., spelling) and three things that I'd really like to see changed (and I am sorry I didn't see it earlier, it's just a big PR):

* Let's use the proper capitalized names of our estimators everywhere (as found in `FEP_ESTIMATORS` and `TI_ESTIMATORS`). The we don't have to upper/lowercase input, and we can just use these names as identifiers. If the user writes `ti` instead of `TI` then we just fail — that's easy and correct and reduces ambiguity and complexity. 
* Add a check for correct estimators at the start of `estimate()`.
* Re-raise exceptions to maintain the original failure information and use `logger.exception` inside exception handlers to log all information.

I am not feeling super-happy about the many `pragma: no cover` on code paths that are taken when errors occur but I also understand that some of these are really quite tedious to test. I'd just encourage as little `no cover` as possible and if there's a reasonable way to test (e.g. for incorrect input) then to do it.

xiki-tempula commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,684 @@
+                        end = states.index(lambda_max)
+                self.logger.info(
+                    f'Stage {stage} is from state {start} to state {end}.')
+                result = delta_f_.iloc[start, end]
```
This one should be fine as the series has been explicitly sorted `sort=True` in `alchemlyb.preprocessing.subsampling.statistical_inefficiency`

orbeckst replied:
Is there any downside to using `.loc`? If we see the code as a best-practice example then shouldn't we propagate these best practices, too?

orbeckst replied:
add a comment that iloc is fine because df is sorted — just what you said above

orbeckst replied:
please add the comment

xiki-tempula commented on docs/workflows/alchemlyb.workflows.ABFE.rst:
```diff
@@ -26,11 +29,15 @@ command. ::
 
 
+See :mod:`~alchemlyb.workflows.ABFE` for the explanation with regard to the
+parameters. The next two sections explains the output of the workflow and a
```
Sorry I wonder what do you mean here? The explanation is linked to `:mod:`~alchemlyb.workflows.ABFE``, which is a link and when the user clicks the link a formatted explanation will be shown.

xiki-tempula commented:
@orbeckst Sorry for the delay. I have addressed the comments. I have removed most of the no cov. I have not changed the iloc as it is not so straightforward in this case as is explained in https://github.com/alchemistry/alchemlyb/issues/203. Using loc instead of iloc in tests are still valid.

orbeckst commented:
Tests fail with
> ERROR collecting src/alchemlyb/tests/test_preprocessing.py
Do you know why?

xiki-tempula commented:
@orbeckst Sorry, I have pushed things to the wrong branch.

orbeckst reviewed (changes requested):
Please see comments, sorry to be brief.

xiki-tempula commented on .github/workflows/ci.yaml:
```diff
@@ -26,7 +26,7 @@ jobs:
     runs-on: ${{ matrix.os }}
     strategy:
-      fail-fast: false
+      fail-fast: true
```
I think it is initially true, then in one of my PR, I set it to false so I could see the fails more clearly but I forget to change it back before the PR is merged. So we want to have this as false?

orbeckst commented:
Yes, let’s do “false” to run all CI.

xiki-tempula commented on src/alchemlyb/convergence/convergence.py:
```diff
@@ -51,16 +52,13 @@ def forward_backward_convergence(df_list, estimator='mbar', num=10):
     logger.info('Check data availability.')
 
-    if estimator.lower() == 'mbar':
-        logger.info('Use MBAR estimator for convergence analysis.')
-        estimator_fit = MBAR().fit
-    elif estimator.lower() == 'autombar':
+    if estimator == 'MBAR':
         logger.info('Use AutoMBAR estimator for convergence analysis.')
-        estimator_fit = AutoMBAR().fit
```
I have made it such that the kwargs are passed to the estimator.

xiki-tempula commented on src/alchemlyb/convergence/convergence.py:
```diff
@@ -51,16 +52,13 @@ def forward_backward_convergence(df_list, estimator='mbar', num=10):
     logger.info('Start convergence analysis.')
     logger.info('Check data availability.')
 
-    if estimator.lower() == 'mbar':
```
@orbeckst I have thought about this and the issue is with AutoMBAR. The AutoMBAR is not upper case so you cannot get it with `getattr(alchemlyb.estimators, estimator)`, the solution that I purposed is to use AutoMBAR to replace the MBAR. For pymbar >= 4, this would also allow us to deprecate AutoMBAR easily.

xiki-tempula commented:
@orbeckst Thanks for the review. I think the major issue is with AutoMBAR and the case issue.

To enforce uppercase, we could change the current AutoMBAR to AUTOMBAR. However, this would mean a change of the API.

Or the easiest way would be use AutoMBAR automatically when the user chooses MBAR, this would allow the removal of AutoMBAR more easily later on without changing the interface. 

When method is given as a keyword, AutoMBAR gives the same behaviour as MBAR, it is only when method is None, AutoMBAR will have different behaviour, so I don't think using AutoMBAR to replace MBAR would be very problematic.

xiki-tempula commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,691 @@
+            plt.close(fig)
+
+        if forwrev is not None:
+            ax = self.check_convergence(forwrev, estimator='MBAR',
```
I think in this case, the user could just go down the manual approach.

```
    >>> workflow = ABFE(units='kcal/mol', software='Gromacs', dir=dir,
    >>>                 prefix='dhdl', suffix='xvg', T=298, outdirectory='./')
    >>> workflow.run(skiptime=10, uncorr='dhdl', threshold=50,
    >>>              methods=('mbar', 'bar', 'ti'), overlap='O_MBAR.pdf',
    >>>              breakdown=True)
    >>> workflow.check_convergence(10, estimator='MBAR', dF_t='dF_t.pdf')
```

orbeckst replied:
Well, they could. But what if I actually don't want to run MBAR on my data, e.g.
```python
workflow.run(..., methods=["TI"], ...)
```
so it would be confusing if the convergence analysis was done with a method that I explicitly excluded. I am not saying that this is a very likely case but we don't know which additional estimators we will have. The proposed change `estimators[0]` provides at least some robustness and agency to the user at little cost.

xiki-tempula replied:
I have been thinking about this for some time.

With regard to `estimators[0]`, this is a method that gives better robustness. In some sense, it gives the user agency as it allows the user to choose the method for the convergence. However, in my experience, selecting the first method in methods is not common sense and might be error-prone if BAR is the first estimator.

I think a defensive approach would be that by default convergence analysis is not performed, as it can be quite time-consuming, so `run()` will not trigger convergence analysis.

The user has to do `run(forwrev=10)` to manually trigger the convergence analysis. In the doc for `forwrev`, it will is written that
```
            By default, 'MBAR'
            estimator will be used for convergence analysis. If the dataset
            does not contain u_nk, please run
            meth:`~alchemlyb.workflows.ABFE.check_convergence` manually
            with estimator='TI'.
```
If there is no u_nk, `check_convergence` will give error
```
                    msg = f"u_nk is needed for the f{estimator} estimator. " \
                          f"If the dataset only has dHdl, " \
                          f"run ABFE.check_convergence(estimator='TI') to " \
                          f"use a TI estimator."
                    self.logger.error(msg)
                    raise ValueError(msg)
```
which tells the user that they need to run `check_convergence` to give a different estimator.

orbeckst replied:
I am ok with making convergence analysis opt-in as you describe. Just add the appropriate documentation. Add a short explanation _why_ (Auto)MBAR is chosen as the default.

orbeckst commented on src/alchemlyb/convergence/convergence.py:
```diff
@@ -20,10 +23,12 @@ def forward_backward_convergence(df_list, estimator='mbar', num=10):
         List of DataFrame of either dHdl or u_nk.
-    estimator : {'mbar', 'bar', 'ti', 'autombar'}
+    estimator : {'MBAR', 'BAR', 'TI'}
         Name of the estimators.
```
add deprecation note
```suggestion
        Name of the estimators.
        See the important note below on the use of "MBAR".

        .. deprecated:: 1.0.0
           Lower case input is also accepted until release 2.0.0.
```

orbeckst commented on src/alchemlyb/convergence/convergence.py:
```diff
@@ -43,29 +48,38 @@ def forward_backward_convergence(df_list, estimator='mbar', num=10):
+
     .. versionadded:: 0.6.0
+    .. versionchanged:: 1.0.0
+        The ``estimator`` accepts uppercase input.
```
```suggestion
       The ``estimator`` accepts uppercase input.
       The default for using ``estimator='MBAR'`` was changed from 
       :class:`~alchemlyb.estimators.MBAR` to :class:`~alchemlyb.estimators.AutoMBAR`.
```

orbeckst replied:
Add an entry to CHANGELOG for the change in default behavior.

orbeckst commented on src/alchemlyb/convergence/convergence.py:
```diff
@@ -20,7 +21,7 @@ def forward_backward_convergence(df_list, estimator='mbar', num=10):
     df_list : list
         List of DataFrame of either dHdl or u_nk.
-    estimator : {'mbar', 'bar', 'ti', 'autombar'}
+    estimator : {'MBAR', 'BAR', 'TI'}
```
Ok, let's go with your AutoMBAR solution.

orbeckst commented on src/alchemlyb/convergence/convergence.py:
```diff
@@ -2,11 +2,12 @@
 
-from ..estimators import MBAR, BAR, TI, AutoMBAR
+from ..estimators import BAR, TI
+from ..estimators import AutoMBAR as MBAR
```
We go with your AutoMBAR solution.

orbeckst commented on src/alchemlyb/estimators/mbar_.py:
```diff
@@ -152,6 +152,14 @@ class AutoMBAR(MBAR):
+    ----------
+
+    method : str, optional, default=None
+        The optimization routine to use. This parameter is default to None.
```
```suggestion
        The optimization routine to use. This parameter defaults to ``None``.
```

orbeckst commented on src/alchemlyb/estimators/mbar_.py:
```diff
@@ -152,6 +152,14 @@ class AutoMBAR(MBAR):
+
+    method : str, optional, default=None
+        The optimization routine to use. This parameter is default to None.
+        When a specific method is set, AutoMBAR will behave in the same ways
```
```suggestion
        When a specific method is set, AutoMBAR will behave in the same way
```

orbeckst commented on src/alchemlyb/estimators/mbar_.py:
```diff
@@ -163,30 +171,35 @@ class AutoMBAR(MBAR):
 
     .. versionadded:: 0.6.0
+    .. versionchanged:: 1.0.0
+        The AutoMBAR accepts method argument.
```
```suggestion
       AutoMBAR accepts the `method` argument.
```

orbeckst commented on src/alchemlyb/estimators/mbar_.py:
```diff
@@ -152,6 +152,14 @@ class AutoMBAR(MBAR):
+    method : str, optional, default=None
+        The optimization routine to use. This parameter is default to None.
+        When a specific method is set, AutoMBAR will behave in the same ways
+        as MBAR.
```
Make this important addition very explicit
```suggestion
        as MBAR.

        .. versionadded:: 1.0.0
        
```

orbeckst replied:
Add entry to CHANGELOG

orbeckst commented on src/alchemlyb/estimators/mbar_.py:
```diff
@@ -163,30 +171,35 @@ class AutoMBAR(MBAR):
 
     def _do_MBAR(self, u_nk, N_k, solver_protocol):
-        self.logger.info('Initialise the automatic routine of the MBAR '
-                         'estimator.')
-        # Try the fastest method first
-        try:
-            self.logger.info('Trying the hybr method.')
-            solver_protocol["method"] = 'hybr'
-            mbar, out = super()._do_MBAR(u_nk, N_k, solver_protocol)
-        except pymbar.utils.ParameterError:
+        if solver_protocol["method"] is None:
+            self.logger.info('Initialise the automatic routine of the MBAR '
+                             'estimator.')
+            # Try the fastest method first
             try:
-                self.logger.info('Trying the adaptive method.')
-                solver_protocol["method"] = 'adaptive'
+                self.logger.info('Trying the hybr method.')
+                solver_protocol["method"] = 'hybr'
                 mbar, out = super()._do_MBAR(u_nk, N_k, solver_protocol)
             except pymbar.utils.ParameterError:
-                self.logger.info('Trying the BFGS method.')
-                solver_protocol["method"] = 'BFGS'
-                mbar, out = super()._do_MBAR(u_nk, N_k, solver_protocol)
-        return mbar, out
+                try:
+                    self.logger.info('Trying the adaptive method.')
+                    solver_protocol["method"] = 'adaptive'
+                    mbar, out = super()._do_MBAR(u_nk, N_k, solver_protocol)
+                except pymbar.utils.ParameterError:
+                    self.logger.info('Trying the BFGS method.')
+                    solver_protocol["method"] = 'BFGS'
+                    mbar, out = super()._do_MBAR(u_nk, N_k, solver_protocol)
+            return mbar, out
+        else:
+            return super()._do_MBAR(u_nk, N_k, solver_protocol)
```
document changes in CHANGELOG; add a versionchanged that the order of methods in AutoMBAR changed

orbeckst replied:
Does this change the tests?

xiki-tempula replied:
No, when `method` is None, nothing is changed. I have added a new test to test the case when `method` is given.

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,691 @@
+from ..parsing import gmx, amber
+from ..preprocessing.subsampling import decorrelate_dhdl, decorrelate_u_nk
+from ..estimators import BAR, TI, FEP_ESTIMATORS, TI_ESTIMATORS
+from ..estimators import AutoMBAR as MBAR
```
We'll go with your suggested approach with AutoMBAR. Documentation is key here!

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,684 @@
+                if self.ignore_warnings:
+                    self.logger.warning(msg)
+                else:
+                    raise ValueError(msg)
```
please address

orbeckst replied:
To be honest, as soon as I see "pragma: no cover" I know that eventually I'll see this pop up as not working as thought... can you at least manually trigger it and confirm that it does want it should be doing?

And it is important to keep the original exception.

xiki-tempula replied:
I have added the test to test this. The original exception is recorded in

```
except Exception as exc: 
    msg = f'Error reading read u_nk from {file}.\n{exc}'
    raise ValueError(msg)
```

where the {exc} in msg is the original exception.

orbeckst replied:
Testing is fantastic!

I still don't like raising `ValueError` here. I'd rather see what really happened _and_ also have access to the actual exception. I also don't see it as a problem with values — it's more likely related to I/O. I just don't see a compelling reason to force it to be a ValueError.

xiki-tempula replied:
Sorry, I cannot figure out how to reraise the same exception. I think one thing that I could do is to use exception Chaining.

So I do

```
            except Exception as exc:
                msg = f'Error reading read u_nk from {file}.'
                if self.ignore_warnings:
                    self.logger.exception(msg +
                        'This exception is being ignored because ignore_warnings=True.')
                else:
                    self.logger.error(msg)
                    raise ValueError(msg) from exc
```
 Note the `raise ValueError(msg) from exc` I will get
```
Traceback (most recent call last):
File "<input>", line 1, in <module>
ValueError: Unable to read u_nk.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
File "/Users/zhiyiwu/GitHub/alchemlyb/src/alchemlyb/workflows/abfe.py", line 122, in read
ValueError: Error reading read u_nk from /Users/zhiyiwu/GitHub/alchemtest/src/alchemtest/gmx/benzene/Coulomb/0000/dhdl.xvg.bz2
```

orbeckst replied:
I thought just a bare `raise` will simply re-raise the exception. Does the following not work
```python
            except Exception as exc:
                msg = f'Error reading read u_nk from {file}.'
                if self.ignore_warnings:
                    self.logger.exception(msg +
                        'This exception is being ignored because ignore_warnings=True.')
                else:
                    self.logger.error(msg)
                    raise
```

orbeckst replied:
Or 
```python
            except Exception as exc:
                msg = f'Error reading read u_nk from {file}.'
                if self.ignore_warnings:
                    self.logger.exception(msg +
                        'This exception is being ignored because ignore_warnings=True.')
                else:
                    self.logger.raise(msg)
                    raise
```
so that the logger includes everything automatically.

xiki-tempula replied:
Thanks for the suggestion. The raise would work but one would lose the error message saying which file gives the problem.
The log file does contain all the information but I think it might be easier to just dump everything into the console so the user doesn't need to open the log file to do the debugging. The exception chaining will preserve the original exception and just append an additional exception at the end which tells the user which file contains issue.

orbeckst replied:
I see the concern that we don't have the filename included. 

What I still don't like is that the actual error that is raised in the end is the `ValueError`.  Here is an example where a `TypeError` is the "inner error":

```python
In [2]: def bad():
   ...:     try:
   ...:         raise TypeError("bad type")
   ...:     except Exception as exc:
   ...:         raise ValueError(f"Oh no {exc} happened!") from exc
   ...:

In [3]: bad()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [2], in bad()
      2 try:
----> 3     raise TypeError("bad type")
      4 except Exception as exc:

TypeError: bad type

The above exception was the direct cause of the following exception:

ValueError                                Traceback (most recent call last)
Input In [3], in <cell line: 1>()
----> 1 bad()

Input In [2], in bad()
      3     raise TypeError("bad type")
      4 except Exception as exc:
----> 5     raise ValueError(f"Oh no {exc} happened!") from exc

ValueError: Oh no bad type happened!
```
and I cannot catch the TypeError
```python
In [5]: try:
   ...:     bad()
   ...: except TypeError:
   ...:     print("Caught the TypeError!")
   ...:
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [2], in bad()
      2 try:
----> 3     raise TypeError("bad type")
      4 except Exception as exc:

TypeError: bad type

The above exception was the direct cause of the following exception:

ValueError                                Traceback (most recent call last)
Input In [5], in <cell line: 1>()
      1 try:
----> 2     bad()
      3 except TypeError:
      4     print("Caught the TypeError!")

Input In [2], in bad()
      3     raise TypeError("bad type")
      4 except Exception as exc:
----> 5     raise ValueError(f"Oh no {exc} happened!") from exc

ValueError: Oh no bad type happened!
```

There's a hacky way to maintain the original exception type:
```python
In [6]: def bad():
   ...:     try:
   ...:         raise TypeError("bad type")
   ...:     except Exception as exc:
   ...:         raise type(exc)(f"Oh no {exc} happened!") from exc
   ...:
```

Maybe there's a better way by building a new exception from `exc` that has the file information appended. Something like
```python
args = (exc.args[0] + f" --- Error occured in file {filename}."  ,) + exc.args[1:]
new_exc = type(exc)(*args)
raise new_exc
```

xiki-tempula replied:
I see your concern now. Sorry I initially didn't see the reason that the original exception needs to be preserved. I have updated the code to do `raise type(exc)(msg)`

orbeckst replied:
`raise type(exc)(msg)` might not be enough information for downstream where exceptions might be processed. I imagine an IOError that comes with an error code — these are part of the exception. Calling code might really want to know if this EACCESS, NOMEM or whatever the codes are. I think by rebuilding the exception as above we can preserve _all_ information.

Sorry for the long discussion.

xiki-tempula replied:
Sorry, the `msg` contains the exception

```
            except Exception as exc:
                msg = f'Error reading read u_nk from {file}.\n{exc}'
                if self.ignore_warnings:
                    self.logger.exception(msg +
                        'This exception is being ignored because ignore_warnings=True.')
                else:
                    self.logger.exception(msg)
                    raise type(exc)(msg)
```

orbeckst replied:
The message is good but cannot be used programmatically. For example, more complex exceptions can be raised
```python
>>> raise IOError(errno.ENOENT, "The file does not exist", "filename.txt")
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
Input In [39], in <cell line: 1>()
----> 1 raise IOError(errno.ENOENT, "The file does not exist", "filename.txt")

FileNotFoundError: [Errno 2] The file does not exist: 'filename.txt'
```
Note that (1) the errno changes the IOError into a FileNotFoundError and that (2) the exception has additional attributes such as errno and filename
```python
>>> exc = IOError(errno.ENOENT, "The file does not exist", "filename.txt")
>>> print(exc.filename, exc.errno)
filename.txt 2
```
all of which get lost when we just use the textual message. 

With the correct exception raised, calling code can trap any IOError and directly figure which file was responsible. I don't know how many people want to do this but why should we limit the possibilities? Transparent exceptions seem to be a good thing to me.
```

xiki-tempula replied:
Thanks for the comment. I have fiddled with this a bit and I think exception chaining could do this in an elegant fashion.
So instead of doing `raise ValueError(msg) from exc`, where `ValueError` is the final exception, I do `raise exc from ValueError(msg)`, in this case, the original exception `exc` is the final exception and the exception that shows which file gives error will be conserved, which also avoids editing the exception message to combine the two.

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,684 @@
+                if self.ignore_warnings:
+                    self.logger.warning(msg)
+                else:
+                    raise ValueError(msg)
```
please address

orbeckst replied:
same as above — testing ❤️ , `ValueError` 👎

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,691 @@
+            states for all methods, including the dH/dlambda curve for TI. Default:
+            ``True``.
+        forwrev : int
+            Plot the free energy change as a function of time in both directions,
+            with the specified number of points in the time plot. The number of time
+            points (an integer) must be provided. Specify as ``None`` will not do
+            the convergence analysis. Default: 10.
```
please address

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,691 @@
+        ----------
+        methods : str or list of str
+            A list of the methods to estimate the free energy with. Default:
+            ['TI', 'BAR', 'MBAR'].
```
use your approach

orbeckst reviewed (changes requested):
I am ok with your approach to use AutoMBAR instead of MBAR, thanks to the documentation that you added and the use of `methods`. I had a bunch of doc changes that you can just accept. Please make sure that CHANGELOG records all behavior changes and additions.  There were a few small things from my last review that weren't addressed so I pinged you on them again. 

It's been a massive PR but it's really almost there. Sorry for my patchy availability to review.

xiki-tempula commented:
@orbeckst Thanks for the review. I have

- changed the change log
- add the explanation to the convergence analysis
- add tests to cover the extract dhdl and n_uk part

orbeckst commented on CHANGES:
```diff
@@ -19,8 +19,13 @@ The rules for this file:
 
 Changes
   - Drop support for py3.7 (Issue #179, PR #214)
+  - forward_backward_convergence will use AutoMBAR as backend when `MBAR` is
+  selected as the estimator (PR #114).
```
Indentation
```suggestion
  - forward_backward_convergence will use AutoMBAR as backend when `MBAR` is
    selected as the estimator (PR #114).
```

orbeckst commented:
Looking really good, the only thing I'd like to change is the ValueError hiding exceptions from the I/O. Otherwise you convinced me to go with your approach or made all the changes.

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,723 @@
+                        'This exception is being ignored because ignore_warnings=True.')
+                else:
+                    self.logger.error(msg)
+                    raise exc from ValueError(msg)
```
Ok, `raise exc from ValueError(msg)` will raise the original error. It will look a bit funny, though, because it will say that the original error was caused by the ValueError that we introduced purely for the sake of adding the information about the file. I read a bit more about exception chaining https://stackoverflow.com/a/24752607/334357 and https://peps.python.org/pep-3134/#explicit-exception-chaining and that changed my mind because I learned that it seems normal to have an API error (currently ValueError) and then ask for the cause in the `apierror.__cause__` attribute.

I still don't like ValueError but I concede that your original solution
```python
raise ValueError from exc
```
was the right approach.

I'd like to raise an `OSError` as our generic error:
```suggestion
                    raise OSError(msg) from exc
```
This is appropriate for I/O and then the cause will make more sense.

orbeckst replied:
Is this exception tested anywhere??

orbeckst commented on src/alchemlyb/workflows/abfe.py:
```diff
@@ -0,0 +1,723 @@
+                        'This exception is being ignored because ignore_warnings=True.')
+                else:
+                    self.logger.error(msg)
+                    raise exc from ValueError(msg)
```
As above
```suggestion
                    raise OSError(msg) from exc
```

(Btw, `IOError` is just `OSError`... so we don't gain anything by using it, it's just more confusing.)

orbeckst replied:
Is this exception tested anywhere??

orbeckst reviewed (changes requested):
I have oscillated back and forth on the exception handling and came down on your suggested solution but using OSError. I will make these changes and see if it works. I am not sure if you're actually testing the exception handling anywhere. (If yes, I'll see it soon... if no, could we add a test?)

orbeckst commented:
@xiki-tempula congratulations — after 1.5 years and almost 300 comments, your ABFE is finally in!!!! 🚀
